### ISSUE TITLE
Add support for Python 3 as expression language with GraalPython

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#2249]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
We currently only support Python 2.7 via Jython, it would be great to have support for Python 3 via GraalPython.
This would also allow using libraries relying on C modules such as `numpy`.

**Describe the solution you'd like**
Add GraalPython as an expression language.

**Describe alternatives you've considered**
I am not aware of any other Python 3 implementation in Java.

**Additional context**
Linked to issue #2248.

**Comment by tfmorris** [comment]
I think Python 3 support would be great, but it's generally better to form requests as questions or requirements rather than answers or specifications.

A couple of things make upgrading to Python 3 a little bit less urgent than for general Python code:
- Jython 2.7 is not end-of-life and continues to receive security updates
- Many Python 3 features have been backported to Jython 2.6 & 2.7, so there's a fair amount of compatibility in Python 2.7
- In the OpenRefine context, Python is used as a lightweight scripting alternative to GREL, not a large scale software development language

For the above reasons, I think we can afford to wait a little bit for Jython 3 or for alternatives like Graal to mature.

**Comment by wetneb** [comment]
Especially because GraalPython is not designed to be a Jython replacement at all: the authors explicitly warn against running GraalPython on a standard JVM. It is designed to run on GraalVM only. It can be run on a standard JVM, but it might be quite a bit slower than Jython in this context (I haven't made benchmarks).

**Comment by magdmartin** [comment]
For reference, there is no plan for a release of Jython 3 in a foreseeable time (jython/jython/issues/24)

**Comment by thadguidry** [comment]
Tim Felgentreff (GraalVM Python contributor) put up [a great article](https://medium.com/graalvm/moving-from-jython-to-graalvm-cf52c4af6106) on the state of things currently and made a followup comment to my experiments in their issue graalvm/graalpython#96  I don't ever see a need for Jython to rise again, since virtualization is becoming that bridge that connects worlds and the technology is getting better at that every year. And this is the stance that GraalVM is basically taking.

**Comment by wetneb** [comment]
Instead of using Python interpreters written in Java (such as Jython or GraalPython), we could also consider using other architectures for integrating Python. For instance, Py4j is used by KNIME (a similar tool) for their Python integration:
https://github.com/knime/knime-python
This is the same tool that is used for pyspark, a Spark interface in Python:
https://dev.to/steadbytes/python-spark-and-the-jvm-an-overview-of-the-pyspark-runtime-architecture-21gg

**Comment by tfmorris** [comment]
Because OpenRefine invokes Python scripts on a per-cell basis it'll be important to evaluate the invocation overhead of any proposed alternatives.

**Comment by thadguidry** [comment]
From [Py4j About page](https://www.py4j.org/about.html):
> As opposed to [Jython](http://www.jython.org/), Py4J does not execute the Python code in the JVM so it does not need to reimplement the Python language and developers can use any libraries supported by their Python interpreter such as libraries written in Cython. Py4J is more focused than [JPype](http://jpype.sourceforge.net/index.html). For example, it does not attempt to link Python threads to Java threads. JPype also uses JNI to communicate with the JVM while Py4J uses plain old sockets, which are more portable in practice.

> In terms of performance, Py4J has a bigger overhead than both of the previous solutions because it relies on sockets, but if performance is critical to your application, accessing Java objects from Python programs might not be the best idea :-) If you have a particular use case that you want to speed up with Py4J, do not hesitate to write on the mailing list or open an issue and we’ll see what we can do.

**Comment by thadguidry** [comment]
@wetneb Looks like Maven artifacts for Graal are coming soon https://github.com/oracle/graal/issues/6852 and they have a comment on the original issue mentioning this as "if all goes well..."

**Comment by chumer** [comment]
@thadguidry we are currently gathering some feedback on this topic in our community chat: https://graalvm.slack.com/archives/CNBFR78F9/p1691241119810719
if you can find some spare minutes to answer my questions there, that would be great!

**Comment by tfmorris** [comment]
@chumer Are these questions so sensitive that they can only be posted in a closed invite-only chat?

**Comment by chumer** [comment]
@tfmorris not at all. Anyone can get access here: https://www.graalvm.org/slack-invitation/

**Comment by tfmorris** [comment]
@chumer Thanks.

For others interested, the referenced "this topic" is primarily about changes  in the GraalVM Languages packaging (https://github.com/oracle/graal/issues/6852). My take is that there are no compatibility issues, which is what the majority of the questions are about,  for OpenRefine since it doesn't yet use GraalPy. Additionally, the new packaging scheme should be easier (ie possible) for us to work with.

**Comment by chumer** [comment]
@tfmorris the questions were also related to the module path requirement. And how hard it is to use it this way for Openrefine. That is something you can potentially answer even if you are not using it yet.

**Comment by thadguidry** [comment]
Maybe later or sooner @tfmorris or yourself could enhance our benchmark tests to give some comparison for simple Python expressions with/without GraalVM to give some performance characteristic differences?

**Comment by thadguidry** [comment]
It would be most awesome to see basic Rscript support even if still somewhat experimental.  Looks like on my Windows it wouldn't work since I read it's only Linux for now? https://www.graalvm.org/latest/reference-manual/r/

**Comment by thadguidry** [comment]
@wetneb You can now update your refine-gython to instead use the new and improved GraalPy Polyglot API which should resolve those final issues and make things much simpler.

Here's the migration instructions from Jython to Python (using Polyglot dependency from GraalPy):
https://github.com/graalvm/graal-languages-demos/tree/main/graalpy/graalpy-jython-guide

Here's the Oracle team's talk on the improvements: https://youtu.be/F8GoDqTtSOE?si=mJ_ncu0MdNjenNiE
In particular the embedded scripting example: https://youtu.be/F8GoDqTtSOE?si=qrDu3C8wcQLkvs0w&t=1579
